src/component_container.jl

```diff
@@ -0,0 +1,71 @@
+"A data structure that acts like a container of components"
+abstract type ComponentContainer <: InfrastructureSystemsContainer end
+
+# For each of these functions, `ComponentContainer` concrete subtypes MUST either implement
+# appropriate methods or accept the default if it exists.
```

Comment:
@daniel-thom these are the only ones I need for `ComponentSelector` and I'm trying to avoid further scope creep. But maybe we also want `iterate_components`, etc. here?

Reply:
Mabye. I'm definitely OK leaving it out.